[MRG] Normalization fix for gbm ensemble feature importances

gforsyth opened the pull request:
#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.
This makes a small change to the ``BaseDecisionTree`` and
``DecisionTreeRegressor``, adding a ``normalize_feature_importances`` ``kwarg``
that defaults to ``True`` (matching current behavior) but is set to ``False``
for GBMs.

Tree-level feature importances are calculated as the proportional reduction of
that tree's root node entropy (see [_tree.pyx](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/tree/_tree.pyx#L1056)) and ensemble-level importances are
averaged across trees.

Because the feature importances are (currently, by default) normalized and then
averaged, feature importances from later stages are overweighted.

GBMs now defer the normalization step until after the per-tree importances have
been calculated (for the ``DecisionTreeRegressor`` used in
``ensemble.gradient_boosting.BaseGradientBoosting``).


#### Any other comments?
h/t to Joel Danke for spotting this and writing up the issue.

gforsyth committed:
Normalization fix for gbm ensemble feature importances

This makes a small change to the ``BaseDecisionTree`` and
``DecisionTreeRegressor``, adding a ``normalize_feature_importances`` ``kwarg``
that defaults to ``True`` (matching current behavior) but is set to ``False``
for GBMs.

Tree-level feature importances are calculated as the proportional reduction of
that tree's root node entropy (see ``_tree.pyx``) and ensemble-level importances are
averaged across trees.

Because the feature importances are (currently, by default) normalized and then
averaged, feature importances from later stages are overweighted.

GBMs now defer the normalization step until after the per-tree importances have
been calculated (for the ``DecisionTreeRegressor`` used in
``ensemble.gradient_boosting.BaseGradientBoosting``).

h/t to Joel Danke for spotting this and writing up the issue.

Signed-off-by: Gil Forsyth <Gilbert.Forsyth@capitalone.com>

jnothman commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -452,6 +452,32 @@ def test_max_feature_regression():
+    sorted_idx = np.argsort(feature_importance)
+
+    feature_order = ['CHAS', 'ZN', 'RAD', 'INDUS', 'B', 'AGE', 'TAX',
+                     'PTRATIO', 'CRIM', 'NOX', 'DIS', 'RM', 'LSTAT']
```
Is this expected order known from elsewhere, or something that you've calculated after the fix?

gforsyth replied:
This is calculated after the fix and included as a non-regression test.

jnothman commented on sklearn/ensemble/gradient_boosting.py:
```diff
@@ -1234,6 +1235,7 @@ def feature_importances_(self):
             total_sum += stage_sum
 
         importances = total_sum / len(self.estimators_)
+        importances /= importances.sum()
```
Please clarify how this relates to the new parameter

gforsyth replied:
The previous behavior normalized by the sum of importances, but did this on a per-tree basis before adding up all of the importances across trees.  The new parameter stops the normalization on the per-tree basis so it is added back in here.

The general idea is to switch from a `normalize then sum` workflow to a `sum then normalize` in order to avoid overweighting features that contribute less to total gini importance.

glemaitre replied:
But when are we using `normalize_feature_importances` actually. It seems that it is never used.

gforsyth replied:
It's never being used in GBM, but it is set `True` (by default) when used by Random Forests.

glemaitre replied:
I got that but I don't see a statement:

```python
if normalize_feature_importances:
    importance /= ...
```

gforsyth replied:
Ahhh, gotcha.  Ok, I'll add that in.  Thanks!

gforsyth replied:
Actually -- sorry, I think I misunderstood.  

It's only being used in the call to instantiate the `DecisionTreeRegressor` without adding in a totally separate class.

glemaitre replied:
Oh sorry I see. The computation of the importance is in the cython file, right.

gforsyth replied:
Right, so for GBMs only, we skip the normalization part of the computation in the cython file and then do the normalization here (which is reasonably cheap since it's just the single numpy operation no matter the number of trees)

jnothman reviewed (commented):
Thanks for the fully-fledged pr. You don't seem to test the effect of changing the new parameter as such

gforsyth commented:
In re: the effect of changing the parameter, it is going to change the feature importance orders -- for the example here: http://scikit-learn.org/stable/auto_examples/ensemble/plot_gradient_boosting_regression.html
both the relative importances and the order of the features will be different.

gforsyth commented:
Thanks for looking it over!

jnothman commented:
okay. my main question (in some ignorance) is: should we be making this an
option, and if so, only in gbm? Or is it a bug that should be fixed? If
it's not an outright bug, then we should be maintaining a
backwards-compatible default.

gforsyth commented:
It's a bug that should be fixed.

gforsyth commented:
Hey @jnothman -- just a gentle ping on this.  Is there anything I can add to help get this merged in?

jnothman commented:
Thanks for the ping.

If it's definitely a bug, is there any benefit to providing an option to switch it? On the other hand, how much user code to we break by changing it outright? (Perhaps it's not so bad if we're only breaking feature importances and not prediction??)

I'm basically unsure about the correct way to go on this, and would appreciate others' opinion and expertise on gradient boosting.

ahmadia commented:
Sorry for the wall of text, here's an academic reference for the issue we're discussing here:

In Friedman’s “Greedy Function Approximation” in the Annals of Statistics, 2001, https://projecteuclid.org/download/pdf_1/euclid.aos/1013203451, the relative importance of input variables is described in section 8.1. Equation 44 (from Breiman, Friedman, Olshen & Stone, 1983) shows that a feature’s relative importance in a tree is the total improvement in squared error over all nodes splitting on that feature — not normalized or proportional — with equation 45 taking the average over all trees of the sum (again, not the average over proportions).

The PR is “canonical” and consistent with other implementations, e.g., R’s gbm (and others).

(Hat tip again to Joel Danke at Capital One for originally spotting this and helping Gil and I with the context we're providing here).

gforsyth commented:
Hey @jnothman -- thanks again for sticking with us on this.  Do you have any suggestions on who among the maintainers has the requisite expertise on gradient boosting to look this over?

jnothman commented:
@glemaitre, if you can chime in on whether you think this should be treated as a bug fix or an option?

glemaitre commented:
This look like a bug from what I can read in Friedman's paper.

glemaitre commented:
I don't think that the fix is the best one (API related). Introducing a parameter in the tree which is not related to the tree or not useful to user is a drawback (we already have a lot of parameter to understand in tree).

However, I think that we could call `compute_features_imporance` directly in the GradientBoosting.

glemaitre commented:
So modifying this function:

https://github.com/scikit-learn/scikit-learn/blob/066b5019875747f7ab1d27a05e4c031446f0ea0d/sklearn/ensemble/gradient_boosting.py#L1219-L1237

glemaitre commented:
So for me the fix should be:

```python
stage_sum = sum(tree.tree_.compute_feature_importances_(normalize=False) 
                for tree in stage) / len(stage) 
```

gforsyth committed:
Revert API changes, move fixed feature importances to GBM BaseClass

gforsyth commented:
Hey @glemaitre -- thanks for the review!  I've reverted the API changes and added in the feature importance fix as you suggested.  If it looks good to you I'll make sure CI passes and then squash down the changes and fix the original commit message to match the actual fix implemented.

glemaitre commented:
> then squash down the changes and fix the original commit message to match the actual fix implemented.

Do not squash. It will mess up the comments on GitHub. We are going to squash with the GitHub interface and take care about the history at merging time ;)

glemaitre commented on doc/modules/ensemble.rst:
```diff
@@ -782,7 +782,7 @@ accessed via the ``feature_importances_`` property::
     ...     max_depth=1, random_state=0).fit(X, y)
     >>> clf.feature_importances_  # doctest: +ELLIPSIS
-    array([0.11, 0.1 , 0.11, ...
+    array([0.10684213, 0.10461707, 0.11265447, ...
```
You might want to use the ellipsis to avoid precision issue:

```python
array([0.10..., 0.10..., 0.11..., ...
```

glemaitre commented on doc/whats_new/v0.20.rst:
```diff
@@ -378,6 +379,12 @@ Classifiers and regressors
 
+- Fixed a bug in :class:`ensemble.gradient_boosting.BaseGradientBoosting` to
+  have feature importances summed and then normalized, rather than normalizing
+  on a per-tree basis. The previous behavior over-weighted the GINI importance
```
Gini

glemaitre commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -452,6 +452,32 @@ def test_max_feature_regression():
+    feature_order = ['CHAS', 'ZN', 'RAD', 'INDUS', 'B', 'AGE', 'TAX',
+                     'PTRATIO', 'CRIM', 'NOX', 'DIS', 'RM', 'LSTAT']
+
+    assert_equal(list(boston.feature_names[sorted_idx]), feature_order)
```
use assert instead of assert equal.

use `tolist()` instead of the `list` constructor

`assert boston.feature_names[sorted_idx].tolist() == ordered_features`

glemaitre commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -452,6 +452,32 @@ def test_max_feature_regression():
+                                  feature_importance.max())
+    sorted_idx = np.argsort(feature_importance)
+
+    feature_order = ['CHAS', 'ZN', 'RAD', 'INDUS', 'B', 'AGE', 'TAX',
```
`ordered_features` is a better name I think

glemaitre commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -452,6 +452,32 @@ def test_max_feature_regression():
+    # over-weighting features from later stages
+    boston = datasets.load_boston()
+    X, y = shuffle(boston.data, boston.target, random_state=13)
+    X = X.astype(np.float32)
```
Why do we convert here?

glemaitre commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -452,6 +452,32 @@ def test_max_feature_regression():
+    boston = datasets.load_boston()
+    X, y = shuffle(boston.data, boston.target, random_state=13)
+    X = X.astype(np.float32)
+    offset = int(X.shape[0] * 0.9)
```
Use the proper function `test_train_split` from the `model_selection` module

glemaitre commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -452,6 +452,32 @@ def test_max_feature_regression():
 
 
+def test_feature_importance_regression():
+    # Test to make sure GINI importance is calculated correctly
```
Gini

gforsyth committed:
Fix typo in method name

glemaitre reviewed (changes requested):
Some nitpicks.

I am not sure about the regression test.
@jnothman do you have any other way to ensure the feature importance order.

gforsyth committed:
Style changes, elide precision on doctest

gforsyth commented:
Hey @glemaitre -- I've made the changes you requested.  The style stuff in the regression test (and the non-standard train-test split) was pulled from [this][0] example in the docs.  

I'm happy to also update the example code but that seems like it should be in a separate PR...?

[0]: http://scikit-learn.org/stable/auto_examples/ensemble/plot_gradient_boosting_regression.html

glemaitre commented on sklearn/ensemble/gradient_boosting.py:
```diff
@@ -1229,11 +1229,12 @@ def feature_importances_(self):
         total_sum = np.zeros((self.n_features_, ), dtype=np.float64)
         for stage in self.estimators_:
-            stage_sum = sum(tree.feature_importances_
+            stage_sum = sum(tree.tree_.compute_feature_importances(normalize=False)
```
The line is too long for PEP8

glemaitre commented:
> I'm happy to also update the example code but that seems like it should be in a separate PR...?

Yes in a separate PR that we can review in parallel. It will be merged much faster since that this is not controversial.

glemaitre commented:
The test failure in python 2 is however an issue. I might mean that the test if flaky but it should interesting what is the difference between the results in different environment. It could be something fishy.

glemaitre commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -452,6 +452,31 @@ def test_max_feature_regression():
+    # Make sure that results are summed, then normalized to avoid
+    # over-weighting features from later stages
+    boston = datasets.load_boston()
+    X, y = shuffle(boston.data, boston.target, random_state=13)
```
I don't think that you need to shuffle.
You could shuffle directly inside the train_test_split

gforsyth committed:
Fix flake failure and remove extra shuffle

jnothman commented on doc/whats_new/v0.20.rst:
```diff
@@ -29,6 +29,7 @@ random sampling procedures.
 - :class:`neural_network.MLPRegressor` (bug fix)
 - :class:`neural_network.MLPClassifier` (bug fix)
 - :class:`neural_network.BaseMultilayerPerceptron` (bug fix)
+- :class:`ensemble.gradient_boosting.BaseGradientBoosting` (bug fix)
```
I think we should probably list the public classifier/regressor rather than Base. We should also note that it only affects feature importances

jnothman reviewed (approved):
Otherwise lgtm

gforsyth committed:
Fix flake error and update news entry

glemaitre commented:
I saw an example of feature importance from Elements of Statistical Learning which is done on california housing. I would like to reproduce it to check if we can use it as testing.                                                                                                                                                                         Sent from my phone - sorry to be brief and potential misspell.

glemaitre commented:
So here come what I would consider as a test:

```python
import numpy as np                                                                   
from sklearn.datasets import fetch_california_housing                                
from sklearn.ensemble import GradientBoostingRegressor                               
from sklearn.ensemble import RandomForestRegressor                                   
from sklearn.model_selection import train_test_split                                 
                                                                                     
california = fetch_california_housing()                                              
X, y = california.data, california.target                                            
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)            
                                                                                     
# Gradient boosting                                                                  
reg = GradientBoostingRegressor(loss='huber', learning_rate=0.1,                     
                                max_leaf_nodes=6, n_estimators=800,                  
                                random_state=0)                                      
reg.fit(X_train, y_train)                                                            
sorted_idx = np.argsort(reg.feature_importances_)[::-1]                              
print('Gradient boosting')                                                           
print([california.feature_names[s] for s in sorted_idx])                             
print(reg.feature_importances_[sorted_idx])                                          
                                                                                     
# Random Forest                                                                      
reg = RandomForestRegressor(n_estimators=100, n_jobs=-1, min_samples_leaf=5,         
                            random_state=0)                                          
reg.fit(X_train, y_train)                                                            
sorted_idx = np.argsort(reg.feature_importances_)[::-1]                              
print('Random forest')                                                               
print([california.feature_names[s] for s in sorted_idx])                             
print(reg.feature_importances_[sorted_idx])
```

### Results in the master branch

```
Gradient boosting
['Latitude', 'Longitude', 'MedInc', 'AveOccup', 'AveRooms', 'AveBedrms', 'HouseAge', 'Population']
[0.19931333 0.19822656 0.14355326 0.13547336 0.10937617 0.09139491
 0.06798544 0.05467696]
Random forest
['MedInc', 'AveOccup', 'Latitude', 'Longitude', 'HouseAge', 'AveRooms', 'Population', 'AveBedrms']
[0.56970374 0.13632459 0.08061255 0.08030935 0.05368477 0.03888769
 0.02031201 0.02016529]
```
![img2](https://user-images.githubusercontent.com/7454015/41667869-4befb9ee-74ae-11e8-83fe-9a991fa57efd.png)

### Results in the current PR:

```
Gradient boosting
['MedInc', 'Longitude', 'AveOccup', 'Latitude', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population']
[0.58500419 0.12820293 0.11039006 0.10886243 0.03395709 0.02155092
 0.00812921 0.00390317]
Random forest
['MedInc', 'AveOccup', 'Latitude', 'Longitude', 'HouseAge', 'AveRooms', 'Population', 'AveBedrms']
[0.56970374 0.13632459 0.08061255 0.08030935 0.05368477 0.03888769
 0.02031201 0.02016529]
```

![img1](https://user-images.githubusercontent.com/7454015/41667406-3360354e-74ad-11e8-844e-447bd4021d70.png)

### Results mentioned in Hastie et al.

![img3](https://user-images.githubusercontent.com/7454015/41668002-a1fad710-74ae-11e8-8c17-1861bec863e3.png)

glemaitre commented:
So it seems that this is a good fix even if the relative importance is still not the same than in the book but the trend is the same.

So I would rework the test based on the snippet that I showed (we might not need 800 estimators). As some test conditions, I would check that MedInc > Longitude > HouseAge > Population.

And I would add a comment at the beginning of the text referring to the experiment in Hastie et al. p.373

glemaitre commented:
@jnothman Do you agree with those?

jnothman commented:
I don't know how brittle that ordering is, but it's certainly comforting that this pr brings us closer. Yes it seems an appropriate test

gforsyth committed:
Replace regression test with example from ESL

Replacing regression test with a GBM example from the Elements of
Statistical Learning, pg. 373.

ogrisel committed:
Faster non-regression test

ahmadia commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -12,7 +12,7 @@
 
 from sklearn import datasets
```
Is this line still necessary?

ogrisel replied:
Yes it's used elsewhere in the test file.

ahmadia commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -453,28 +453,27 @@ def test_max_feature_regression():
-    assert boston.feature_names[sorted_idx].tolist() == ordered_features
+    """Test that Gini importance is calculated correctly
+
+    This test is follows an example from [1] (pg. 373).
```
-> The test that follows is an example...

ahmadia commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -453,28 +453,27 @@ def test_max_feature_regression():
+    [1]: Friedman, J., Hastie, T., & Tibshirani, R. (2001). The elements of
+    statistical learning. New York: Springer series in statistics.
+    """
+    california = fetch_california_housing()
```
Doesn't PEP8 want a newline here?

ogrisel replied:
No, I don't think so.

ahmadia commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -453,28 +453,27 @@ def test_max_feature_regression():
+    X, y = california.data, california.target
+    X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)
+
+    # Gradient boosting
```
Is this comment REALLY necessary?

ogrisel replied:
I will clean it up when merging but let me wait for the CI first. appveyor is already saturated :)

glemaitre replied:
I made the cleanup.

ahmadia commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -453,28 +453,27 @@ def test_max_feature_regression():
+    sorted_idx = np.argsort(reg.feature_importances_)[::-1]
+    feature_names_sorted = [california.feature_names[s] for s in sorted_idx]
+
+    assert feature_names_sorted[:4] == ['MedInc', 'Longitude',
```
It might be worth testing the bottom 2-3 as well if they are stable. But at this point, I'm really happy with where this PR is.

ogrisel commented:
Thanks for the non-regression test @gforsyth . I found it a bit too slow. I pushed a change to relax it a bit while checking that I could still reproduce the issue on master with only 100 trees (instead of 200).

ogrisel commented:
BTW @gforsyth, remark for later: you should not use your master branch to submit PRs, otherwise you cannot work on several PRs in parallel.

ogrisel reviewed (approved):
LGTM if CI is still green with the last commit.

ogrisel commented:
@ahmadia @gforsyth in the plots reported by @glemaitre there seems to be a difference in the relative feature importances w.r.t. the plot of ESLII. It would be great to try to understand the source of that remaining discrepancy (although I think only the order really matters in practice). I you have sparse cycles feel free to open another issue with the results of your investigations.

gforsyth commented:
> I found it a bit too slow. I pushed a change to relax it a bit while checking that I could still reproduce the issue on master with only 100 trees (instead of 200).

Sounds good to me @ogrisel -- thanks for the fix.



> you should not use your master branch to submit PRs, otherwise you cannot work on several PRs in parallel.

can't I just open PRs from arbitrary branches?

ogrisel commented:
> can't I just open PRs from arbitrary branches?

You can but it's confusing to have a branch named "master" that does not actually track the content of the upstream master branch.

jnothman reviewed (commented):
Thanks and good work, everyone.

glemaitre commented:
> @ahmadia @gforsyth in the plots reported by @glemaitre there seems to be a difference in the relative feature importances w.r.t. the plot of ESLII. It would be great to try to understand the source of that remaining discrepancy (although I think only the order really matters in practice). I you have sparse cycles feel free to open another issue with the results of your investigations.

I would even assume that this something shared by the GradientBoosting and RandomForest. I would check if there is not something fishy with some normalization when computing the the Gini Importance descending each tree.

gforsyth commented:
CI passed!  Thanks for the reviews and help, all!

glemaitre committed:
nitpicks
